<a href="https://colab.research.google.com/github/bacchus00/tareas-ds/blob/main/Tarea_1_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Melanie Marquez, Franco Zautzik

**Fecha límite de entrega 📆:** Martes 29 de Marzo.

**Tiempo estimado de dedicación:** 5 hrs




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 5 lineas** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones.**



` ` 

Es compleja por que el lenguaje humano es ambiguo en cuanto a sus reglas y adicionalmente esta siempre cambiando. Tambien tiene la cualidad de ser Sparse, puesto que se pueden combinar de infinitas formas las palabras para generar una oracion que tenga sentido, y es composicional (Compositionality) que significa que el significado de una oracion es mas que solo el significado individual de cada palabra.
` ` 

**Pregunta 2 (0.3 puntos): ¿Cuál es la diferencia entre Natural Language Processing (NLP) y Computational Linguistics?**

` ` 

NLP es una diciplina aplicada que se enfoca en resolver tasks bien de definidas que involucran el lenguaje humano mientras que computational linguistics busca, con ayuda de la computacion, entender de que manera funciona el lenguaje humano.

` ` 

**Pregunta 3 (0.3 puntos): ¿Qué es el Machine Learning supersivado? Donde podría aplicar este tipo de Aprendizaje**

` ` 

El machine learning supervisado es un tipo de ML que en base a un data set previamente anotado de pares input, output aprende a mapear nuevos inputs a su ouput correspondiente. Un ejemplo de aplicacion en hacer un clasificador de imagenes que distinga entres si son perros o gatos.

` ` 

**Pregunta 4 (0.3 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.**

` ` 

El Machine Learning tradicional utiliza una representacion inicial del input y en base a la data de entrenamiento previamente anotada aprende la relacion entre input y output,
al Deep Learning, encambio, se le entega el input en un formato "raw" y este por si solo aprende formas de representar la data y patrones en los datos.

` ` 

 **Pregunta 5 (0.3 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 

Puede ocupar la representacion bag of words de los documentos llevando estos a si a un espacio vectorial. Luego se puede utilizar la similitud coseno para medir la similitud entre documentos en base a su representacion vectorial y rankear. El problema con esta forma es que las bag of words obvian completamente el orden que tienen las palabras dentro de un documento.

` ` 

**Pregunta 6 (0.3 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.**

` ` 

El jefe se refiere a las Stop Words, palabras que son muy frecuentes en un vocabulario, cosa que comprende los articulos, nombres, pronombres, preposiciones, conjunciones, etc... En analisis de sentimiento eliminar todas las stop words puede ser problematico por que podria eliminar la palabra "not" cambiando completamente el sentimiento de alguna oracion. Por eljemplo "I do not like bananas" -> "like bananas"

` ` 

**Pregunta 7 (0.3 puntos) Nombre las ventajas y desventajas de utilizar un modelo Bag of Words y TFIDF**

` ` 

Desventajas:

*   El modelo bag of words pierde completamente el orden en el que estan puestas las palabra.
*   No puede capturar la semantica de un documento solo su lexico

Ventajas:


*   Es facil de calcular
*   Elemento de la listaPemite tener la nocion de cuan frecuentes son las palabras en un documento (TF)
*   Toma en consideracion cuan relevante es una palabra en un documento considerando su aparicion en los demas documentos (IDF) 

` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Typos*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [ ]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [ ]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [ ]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

In [ ]:
try:
    from google.colab import drive

    drive.mount("/content/drive")
    path = "/content/drive/MyDrive/marcianeke.txt"
except: 
    print('Ignorando conexión drive-colab')

Mounted at /content/drive


` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver que tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [ ]:
text = codecs.open(path, 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [ ]:
def get_tokens(text):
  ### Inicio del código ###
  return re.findall(r"[A-Za-záéíóú\-']+|[.,()]",text,re.UNICODE)
  ### Fín del código ###

In [ ]:
tokens = get_tokens(text)
tokens

['Brr',
 'Marcianeke',
 "Vamo'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata-ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 "quiere'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata-ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 "quiere'",
 'mezclar',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',
 'pide',
 'yo',
 'lo',
 'traje',
 'No',
 'visto',
 'de',
 'traje',
 'Puro',
 'corte',
 'calle',
 ',',
 '

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

El apostrofe " ' " y el guion "-" son considerados parte de las palabras en el tokenizador creado, en el de nltk en cambio son separados de las palabras de las cuales forma parte. Por otro lado el punto ".", coma "," y los parentesis "(" ")" son considerados como tokens por si solos en el tokenizador creado y el nltk 

In [ ]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [ ]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los typos de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [ ]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  tokens = []
  for doc in dataset:
    tokens.extend(get_tokens(doc))
  return list(set(tokens))
  ### Aquí termina tu código ###

***Test:***

In [ ]:
vocab = get_vocab(dataset)
vocab

['my',
 'like',
 'programming',
 'languages',
 'Spanish',
 'language',
 'is',
 'favorite',
 'human',
 'I']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

**Reglas:** Estan hechas ad hoc al corpus

*   En cuanto a las stopwords en la funcion remove_stopwords hay una lista harcodeada de stopwords, esta no es exauhstiva (como lo descrito en la **Pregunta 6**) y esta hecha especificamente para el corpus dado. 
*   En cuanto a al stemming se ocupo la reglA 's' -> '' del algoritmo de porter puesto que el la unica que aplica para el corpus.

In [ ]:
def remove_stopwords(vocab):
  stopwords = ["is", "my", "I"]
  for stopword in stopwords:
    if stopword in vocab:
      vocab.remove(stopword)
  return vocab

def stemming(token):
  if token[-1] == "s":
    return token[:-1]
  else:
    return token

In [ ]:
def pre_processing(vocab): 
  ### Aquí inicia tu código ###
  vocab = remove_stopwords(vocab)
  vocab = list(set(map(stemming, vocab)))
  return vocab
  ### Aquí termina tu código ###

In [ ]:
vocab = pre_processing(vocab)
vocab

['like', 'programming', 'Spanish', 'language', 'favorite', 'human']


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [ ]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [ ]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  doc_names = ["d0","d1","d2","d3","d4","d5"]
  vocab = get_vocab(dataset)
  df = pd.DataFrame(index=doc_names, columns=vocab)
  print(df["perro"]["d0"])
  for i in range(len(dataset)):
    for j in range(len(vocab)):
      df[vocab[j]][doc_names[i]] = get_tokens(dataset[i]).count(vocab[j])
  return df
  ### Aquí termina tu código ###

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

nan


después despierta empieza maullar  y comida come gato El duerme  a ladra  \
d0       1         0       0       0  1      1    1    0  1      1  0     0   
d1       1         1       1       0  1      0    0    0  1      0  1     0   
d2       1         0       0       0  1      1    1    0  1      0  0     1   
d3       1         0       0       0  1      1    1    1  1      1  0     0   
d4       1         1       1       1  1      0    0    1  1      0  1     0   
d5       1         0       0       0  1      1    1    1  1      0  0     0   

   maulla la se ladrar perro  
d0      0  1  2      0     1  
d1      0  0  1      1     1  
d2      0  1  1      0     1  
d3      0  1  2      0     0  
d4      0  0  1      0     0  
d5      1  1  1      0     0

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    doc_names = ["d0","d1","d2","d3","d4","d5"]
    vocab = dataset_bow.columns
    max_tf = dataset_bow.max(axis=1)
    for i in range(len(dataset_bow.index)):
      for j in range(len(dataset_bow.columns)):
        dataset_bow[vocab[j]][doc_names[i]] = dataset_bow[vocab[j]][doc_names[i]]/max_tf[i]
        
    return dataset_bow
    ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

después despierta empieza maullar    y comida come gato   El duerme    a  \
d0     0.5       0.0     0.0     0.0  0.5    0.5  0.5  0.0  0.5    0.5  0.0   
d1     1.0       1.0     1.0     0.0  1.0    0.0  0.0  0.0  1.0    0.0  1.0   
d2     1.0       0.0     0.0     0.0  1.0    1.0  1.0  0.0  1.0    0.0  0.0   
d3     0.5       0.0     0.0     0.0  0.5    0.5  0.5  0.5  0.5    0.5  0.0   
d4     1.0       1.0     1.0     1.0  1.0    0.0  0.0  1.0  1.0    0.0  1.0   
d5     1.0       0.0     0.0     0.0  1.0    1.0  1.0  1.0  1.0    0.0  0.0   

   ladra maulla   la   se ladrar perro  
d0   0.0    0.0  0.5  1.0    0.0   0.5  
d1   0.0    0.0  0.0  1.0    1.0   1.0  
d2   1.0    0.0  1.0  1.0    0.0   1.0  
d3   0.0    0.0  0.5  1.0    0.0   0.0  
d4   0.0    0.0  0.0  1.0    0.0   0.0  
d5   0.0    1.0  1.0  1.0    0.0   0.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [ ]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    N = len(dataset_bow.index)
    vocab = dataset_bow.columns
    doc_names = dataset_bow.index
    dic = {}
    for j in range(len(dataset_bow.columns)):
      suma = 0
      for i in range(len(dataset_bow.index)):
        if dataset_bow[vocab[j]][doc_names[i]] > 0:
          suma+=1
      dic[vocab[j]] = np.log10(N/suma)
    return dic
    pass
    ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    doc_names = tf.index
    vocab = tf.columns
    for i in range(len(tf.index)):
      for j in range(len(tf.columns)):
        tf[vocab[j]][doc_names[i]] = idf[vocab[j]] * tf[vocab[j]][doc_names[i]]
    return tf
    pass
    ### Aquí termina tu código ### 

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

después despierta   empieza   maullar    y    comida      come      gato  \
d0     0.0       0.0       0.0       0.0  0.0  0.088046  0.088046       0.0   
d1     0.0  0.477121  0.477121       0.0  0.0       0.0       0.0       0.0   
d2     0.0       0.0       0.0       0.0  0.0  0.176091  0.176091       0.0   
d3     0.0       0.0       0.0       0.0  0.0  0.088046  0.088046  0.150515   
d4     0.0  0.477121  0.477121  0.778151  0.0       0.0       0.0   0.30103   
d5     0.0       0.0       0.0       0.0  0.0  0.176091  0.176091   0.30103   

     El    duerme         a     ladra    maulla        la   se    ladrar  \
d0  0.0  0.238561       0.0       0.0       0.0  0.088046  0.0       0.0   
d1  0.0       0.0  0.477121       0.0       0.0       0.0  0.0  0.778151   
d2  0.0       0.0       0.0  0.778151       0.0  0.176091  0.0       0.0   
d3  0.0  0.238561       0.0       0.0       0.0  0.088046  0.0       0.0   
d4  0.0       0.0  0.477121       0.0       0.0       0.0  0.0       0.0   
d5  0.0       0.0       0.0       0.0  0.778151  0.176091  0.0       0.0   

       perro  
d0  0.150515  
d1   0.30103  
d2   0.30103  
d3       0.0  
d4       0.0  
d5       0.0

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  norma = 0
  for i in range(len(v1)):
    norma += (v1[i]*v1[i])
  norma = np.sqrt(norma)
  norma2 = 0
  for i in range(len(v2)):
    norma2 += (v2[i]*v2[i])
  norma2 = np.sqrt(norma2)
  cos = np.dot(v1, v2)/(norma*norma2)
  return cos
  pass
  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


   Los documentos mas similares son aquellos con angulo mas cercano a 0, por lo tanto, con coseno mas cercano a 1. Notese que en la diagonal todos los valores son 1 porque se compara cada documento consigo mismo. Finalmente, los documentos mas similares son d0 con d3, pues su similitud coseno es de 0.8

![gato](https://www.elagoradiario.com/wp-content/uploads/2020/05/Gato-mascarilla.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**